axelrod/strategies/axelrod_second.py

```diff
@@ -771,3 +771,40 @@ def strategy(self, opponent: Player) -> Action:
                 return random_choice(0.5)
         else:
             return C
+
+class WmAdams(Player):
+    """
+    Strategy submitted to Axelrod's second tournament by William Adams (K49R),
+    and came in fifth in that tournament.
+
+    Count the number of opponent defects after their first move, call 
+    `c_defect`.  Defect if c_defect equals 4, 7, or 9.  If c_defect > 9,
+    then defect immediately after opponent defect with probability = 
+    (0.5)^(c_defect-1).  Otherwise cooperate.
+
+    Names:
+
+    - WmAdams: [Axelrod1980b]_
+    """
+
+    name = "WmAdams"
+    classifier = {
+        'memory_depth': float('inf'),
+        'stochastic': True,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def strategy(self, opponent: Player) -> Action:
+        if len(self.history) <=1:
+            return C
+        did_d = np.vectorize(lambda action: float(action == D))
```

Comment:
We keep a running variable of the number of defections in `opponent.defections`. Please use that instead, and subtract one if the first move is a defection. That way is `O(rounds)`, the current implementation is `O(rounds^2)`.

Reply:
Wondered about this.  I can fix in my earlier "Cave" implementation too.

Reply:
If you could that would be great (my bad for missing that in the review :)).